#!/bin/bash
set -e 
#######################################
# load environment variables for LRSDAY
source ./../../env.sh

#######################################
# set project-specific variables
genome="./../06.Supervised_Final_Assembly/SK1.final.fa" # path of the input genome assembly
prefix="SK1" # file name prefix for the output file
centromere_gff3="./../07.Centromere_Annotation/SK1.centromere.gff3" # path of the final centromere annotation gff3 generated by task 07.Centromere_Annotation_for_FinalAssembly, use "" if not exists 
TE_gff3="./../09.TE_Annotation/SK1.TE.gff3" # path of the TE annotation gff3 generated by task 09.TE_Annotation, use "" if not exists
X_element_gff3="./../10.Core_X_Element_Annotation/SK1.X_element.gff3" # path of the X_element annotation gff3 generated by task 10.CoreX_Annotation, use "" if not exists
Y_prime_element_gff3="./../11.Y_Prime_Element_Annotation/SK1.Y_prime_element.gff3" # path of the Y_prime_element annotation gff3 generated by task 11.Yprime_Annotation, use "" if not exists
gene_gff3="./../12.Gene_Orthology_Identification/SK1.updated.gff3" # path of the gene annotation gff3 generated by task 12.GeneOrthology_Identification(better) or 08.Gene_Annotation, use "" if not exists
debug="no" # use "yes" if prefer to keep intermediate files, otherwise use "no".

######################################
# process the pipeline
cat $centromere_gff3 $TE_gff3 $X_element_gff3 $Y_prime_element_gff3 $gene_gff3  >$prefix.concatenated.gff3
perl $LRSDAY_HOME/scripts/sort_gff3.pl -i $prefix.concatenated.gff3 -t $prefix -o $prefix.final.gff3 -r $genome
perl $LRSDAY_HOME/scripts/extract_cds_from_tidy_gff3.pl -r $genome -g $prefix.final.gff3 -o $prefix.final.cds.fa
perl $LRSDAY_HOME/scripts/cds2protein.pl -i $prefix.final.cds.fa -p $prefix.final
cp $genome $prefix.final.fa

# clean up intermediate files
if [[ $debug == "no" ]]
then
    rm $prefix.concatenated.gff3
fi

############################
# checking bash exit status
if [[ $? -eq 0 ]]
then
    echo ""
    echo "LRSDAY message: This bash script has been successfully processed! :)"
    echo ""
    echo ""
    exit 0
fi
############################
